server/src/main/java/org/opensearch/search/GenericSearchExtBuilder.java

```diff
@@ -0,0 +1,160 @@
+ * compatible open source license.
+ */
+package org.opensearch.search;
+
+import org.opensearch.core.ParseField;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParseException;
+import org.opensearch.core.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public final class GenericSearchExtBuilder extends SearchExtBuilder {
+
+    public final static ParseField EXT_BUILDER_NAME = new ParseField("generic_ext");
+
+    private final Object genericObj;
+    private final ValueType valueType;
+
+    enum ValueType {
+        SIMPLE(0),
+        MAP(1),
+        LIST(2);
+
+        private final int value;
+
+        ValueType(int value) {
+            this.value = value;
+        }
+
+        public int getValue() {
+            return value;
+        }
+
+        static ValueType fromInt(int value) {
+            switch (value) {
+                case 0:
+                    return SIMPLE;
+                case 1:
+                    return MAP;
+                case 2:
+                    return LIST;
+                default:
+                    throw new IllegalArgumentException("Unsupported value: " + value);
+            }
+        }
+    }
+
+    public GenericSearchExtBuilder(Object genericObj, ValueType valueType) {
+        this.genericObj = genericObj;
+        this.valueType = valueType;
+    }
+
+    public GenericSearchExtBuilder(StreamInput in) throws IOException {
+        valueType = ValueType.fromInt(in.readInt());
+        switch (valueType) {
+            case SIMPLE:
+                genericObj = in.readGenericValue();
+                break;
+            case MAP:
+                genericObj = in.readMap();
+                break;
+            case LIST:
+                genericObj = in.readList(r -> r.readGenericValue());
+                break;
+            default:
+                throw new IllegalStateException("Unable to construct GenericSearchExtBuilder from incoming stream.");
+        }
+    }
+
+    public static GenericSearchExtBuilder fromXContent(XContentParser parser) throws IOException {
+        // Look at the parser's next token.
+        // If it's START_OBJECT, parse as map, if it's START_ARRAY, parse as list, else
+        // parse as simpleVal
+        XContentParser.Token token = parser.currentToken();
+        ValueType valueType;
+        Object genericObj;
+        if (token == XContentParser.Token.START_OBJECT) {
+            genericObj = parser.map();
+            valueType = ValueType.MAP;
+        } else if (token == XContentParser.Token.START_ARRAY) {
+            genericObj = parser.list();
+            valueType = ValueType.LIST;
+        } else if (token.isValue()) {
+            genericObj = parser.objectText(); // .objectBytes() ??
```

Comment:
I was just checking the difference between `objectText` and `objectBytes`, and they're pretty similar except when dealing with string values:

```
    @Override
    public Object objectText() throws IOException {
        JsonToken currentToken = parser.getCurrentToken();
        if (currentToken == JsonToken.VALUE_STRING) {
            return text();
        } else if (currentToken == JsonToken.VALUE_NUMBER_INT || currentToken == JsonToken.VALUE_NUMBER_FLOAT) {
            return parser.getNumberValue();
        } else if (currentToken == JsonToken.VALUE_TRUE) {
            return Boolean.TRUE;
        } else if (currentToken == JsonToken.VALUE_FALSE) {
            return Boolean.FALSE;
        } else if (currentToken == JsonToken.VALUE_NULL) {
            return null;
        } else {
            return text();
        }
    }
```

```
    @Override
    public Object objectBytes() throws IOException {
        JsonToken currentToken = parser.getCurrentToken();
        if (currentToken == JsonToken.VALUE_STRING) {
            return charBuffer();
        } else if (currentToken == JsonToken.VALUE_NUMBER_INT || currentToken == JsonToken.VALUE_NUMBER_FLOAT) {
            return parser.getNumberValue();
        } else if (currentToken == JsonToken.VALUE_TRUE) {
            return Boolean.TRUE;
        } else if (currentToken == JsonToken.VALUE_FALSE) {
            return Boolean.FALSE;
        } else if (currentToken == JsonToken.VALUE_NULL) {
            return null;
        } else {
            return charBuffer();
        }
    }
```

Given that the only difference is in the string case, I feel like your choice of `objectText` is  the correct one.